src/ndarray/ndarray.cc

```diff
@@ -1017,6 +1017,7 @@ inline void CopyFromToDnsImpl(const NDArray& from, const NDArray& to, RunContext
     // with Copy().
     NDArray tmp_from = from;
     if (tmp_from.IsMKLDNNData()) {
+      // TODO(zhengda) tmp_from should be cached.
```

Comment:
Please create an issue to track this

Reply:
Create a github issue for this?

Reply:
Yes. Ideally, there should be no TODOs at all. But in cases like this it's acceptable to have them, but they must be tracked as otherwise we lose overview.

Reply:
The reason I ask is that I saw TODO in many places in the code. Are all TODOs tracked in github issues? Or is this the new practice we want to push? And is github issue the best place for us to keep track of these TODOs? I remember Sheng has a script that closes issues if they aren't inactive for 30 days or so.

Reply:
In think it's best practice and that we should lead by example :) In the last months I haven't seen any TODO (besides the ones in your MKLDNN PR) checked into MXNet. 

We got rid of the script to automatically close issues. For now, GitHub issues are the right place to track these things.

Reply:
I'm not familiar with this code. But why are all all these operations mutating the array in a function that is supposed to be const?

Reply:
if it doesn’t require const_cast<>, then it’s perfectly legal.

Reply:
const can have a lot of definitions. I don't think it's a bad design. here const means the values of the array are always the same, even though its layout may be changed. if we strictly follow the rule that const function doesn't modify data in an array, the code would become unmanagable. In fact, the data in a const array in MXNet is modified all the time.

Reply:
Why don't we use dynamic dispatch in NDArray for different implementations and storage options? it would make the code more mainable. I predict Chris is going to mention the overhead of a virtual function but I don't think it would be significant also considering the conditional logic in some of the NDArray methods and Python overheads which are magnitudes higher.

Reply:
@cjolivier01  sure it's legal to change data through a pointer from a const method, but having unexpected side effects from a class contract which indicates that a method is const leads to problems, is not particularly considered a good practice when programming: https://stackoverflow.com/questions/763835/are-side-effects-a-good-thing  having experience with functional programming languages helps appreciate controlled side effects.

Reply:
There are many decaffeinated brands out there that are just as tasty as the real thing.

Reply:
Whatever you guys are talking here... I think it's a pretty basic thing that immutable means immutable and that a getter should never modify the underlying data in any way. I have no clue about C++, but I'm quite certain that this constraint is quite clearly defined across the industry. Could we please get back to a productive discussion?

Reply:
@larroy I agree with you that the current implementation of NDArray to handle storages is messy. What you said requires a rewrite of NDArray. @piiswrong and I talked about this, but I don't think this should be in this PR. We need to be very careful when rewriting NDArray, considering NDArray is one of the core components in MXNet.

Reply:
@zheng-da I didn't mean it to be included in this PR. I was just asking. Anyway thanks for your responses. I didn't get the quote about the brands.

Reply:
Open TODO